--- linux-6.1.5-cachyos-DEF/init/Kconfig	2023-01-25 10:03:45.672326190 -0600
+++ linux-6.1.5-cachyos/init/Kconfig	2023-01-25 10:24:39.855767169 -0600
@@ -1497,6 +1497,11 @@
 	  Choosing this option will pass "-O3" to your compiler to optimize
 	  the kernel yet more for performance.
 
+config CC_OPTIMIZE_FOR_PERFORMANCE_OFAST
+  bool "Optimize for just wanting to go fast (-Ofast)"
+  help
+    Choose this option if you just care about speed.
+
 config CC_OPTIMIZE_FOR_SIZE
 	bool "Optimize for size (-Os)"
 	help
