--- linux-6.1.5-cachyos-DEF/drivers/video/aperture.c	2023-01-25 10:03:45.072103848 -0600
+++ linux-6.1.5-cachyos/drivers/video/aperture.c	2023-01-25 13:55:52.277117155 -0600
@@ -332,15 +332,16 @@
 	primary = pdev->resource[PCI_ROM_RESOURCE].flags & IORESOURCE_ROM_SHADOW;
 #endif
 
+  if(primary)
+    sysfb_disable();
+
 	for (bar = 0; bar < PCI_STD_NUM_BARS; ++bar) {
 		if (!(pci_resource_flags(pdev, bar) & IORESOURCE_MEM))
 			continue;
 
 		base = pci_resource_start(pdev, bar);
 		size = pci_resource_len(pdev, bar);
-		ret = aperture_remove_conflicting_devices(base, size, primary, name);
-		if (ret)
-			return ret;
+    aperture_detach_devices(base, size);
 	}
 
 	/*
